### PR TITLE
Regex parser nits

### PR DESCRIPTION
**Pull request by danmoseley**
trivial changes.
1. I ran the real world patterns corpus through the parser and didn't notice anything that could be obviously improved except for one dictionary which was almost always resizing. Set size to eliminate resizing in 90% of cases.
2. Inlined some more low value helpers in RegexParser where I think it makes it clearer.
3. Marked methods readonly where our analyzer instructed it.
4. Some use of IndexOf() per feedback in #88392

### CODE REVIEW

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs [hunk 1]
```diff
@@ -1158,8 +1158,6 @@ private void ScanBlank()
         /// <summary>Scans \-style backreferences and character escapes</summary>
         private RegexNode? ScanBasicBackslash(bool scanOnly)
         {
-            Debug.Assert(_pos < _pattern.Length, "The current reading position must not be at the end of the pattern");
```
Comment (danmoseley) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:1186]: removed these asserts because there'll be a null ref a couple lines down in each case; we don't put asserts in front of every possible null ref.

**Comment by danmoseley** [comment]
@dotnet/area-system-text-regularexpressions would it be possible for one of you to review in Stephen's absence?

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs [hunk 2]
```diff
@@ -2141,22 +2093,6 @@ private void AddAlternate()
...
-            _concatenation!.AddChild(_unit!);
-            _unit = null;
-        }
-
-        /// <summary>Finish the current quantifiable (when a quantifier is found)</summary>
-        private void AddConcatenate(bool lazy, int min, int max)
-        {
-            _concatenation!.AddChild(_unit!.MakeQuantifier(lazy, min, max));
-            _unit = null;
-        }
```
Comment (buyaa-n) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:2158]: NIT: I would keep this methods as is, they have 2 statements and has some comment could be useful for readers, but up to you.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs (hunk 1 above)
Comment (buyaa-n) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:1186]: Not sure I understood the comment, but it looks to me that these are asserting `The current reading position` not a null ref.

**Review by buyaa-n — APPROVED** [review]
Left a few NIT questions, overall LGTM.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs (hunk 1 above)
Comment (danmoseley) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:1186]: I should have said IndexOutOfRangeException

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs [hunk 3]
```diff
@@ -513,33 +516,24 @@ private RegexNode ScanReplacement()
...
                 {
                     if (_pattern[_pos++] == '$')
                     {
-                        RegexNode node = ScanDollar();
-                        _unit = node;
+                        _unit = ScanDollar();
                     }
 
-                    AddConcatenate();
+                    _concatenation.AddChild(_unit!);
```
[the remark this answers is not in the document]
Comment (danmoseley) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:536]: I convinced myself it was redundant but I'm not sure how - I put it back.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs (hunk 2 above)
Comment (danmoseley) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:2158]: Thanks, but I find it clearer

**Comment by danmoseley** [comment]
Thanks for the review

**Comment by danmoseley** [comment]
Test failures unrelated

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs [hunk 4]
```diff
@@ -513,33 +516,25 @@ private RegexNode ScanReplacement()
...
                 int startpos = _pos;
 
-                while (c > 0 && _pattern[_pos] != '$')
-                {
-                    _pos++;
-                    c--;
-                }
+                _pos = _pattern.IndexOf('$', _pos);
+                if (_pos == -1)
+                    _pos = _pattern.Length;
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:525]:
```C#
                if (_pos == -1)
                {
                    _pos = _pattern.Length;
                }
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs [hunk 5]
```diff
@@ -513,33 +516,25 @@ private RegexNode ScanReplacement()
...
 
-                AddConcatenate(startpos, _pos - startpos, isReplacement: true);
+                AddToConcatenate(startpos, _pos - startpos, isReplacement: true);
 
-                if (c > 0)
+                if (_pos < _pattern.Length)
                 {
                     if (_pattern[_pos++] == '$')
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:531]: In what situation will `_pattern[_pos]` here _not_ equal '$'?  Does AddToConcatenate change _pos? If not, aren't we guaranteed that _pattern[_pos] == '$' since if _pos < _pattern.Length that means IndexOf found the '$'?

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs [hunk 6]
```diff
@@ -884,9 +879,9 @@ private RegexNode ScanReplacement()
                                 {
                                     string capname = ScanCapname();
 
-                                    if (IsCaptureName(capname))
+                                    if (_capnames != null && _capnames.ContainsKey(capname))
                                     {
-                                        capnum = CaptureSlotFromName(capname);
+                                        capnum = (int)_capnames![capname]!;
                                     }
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:885]: Could this avoid a secondary hash lookup by instead being:
```C#
if (_capnames?[capname] is int tmpCapnum)
{
    capnum = tmpCapnum;
}
```
?

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs [hunk 7]
```diff
@@ -931,9 +926,9 @@ private RegexNode ScanReplacement()
                                     {
                                         string uncapname = ScanCapname();
 
-                                        if (IsCaptureName(uncapname))
+                                        if (_capnames != null && _capnames.ContainsKey(uncapname))
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:929]: Ditto

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs [hunk 8]
```diff
@@ -991,9 +986,9 @@ private RegexNode ScanReplacement()
                             {
                                 string capname = ScanCapname();
 
-                                if (IsCaptureName(capname) && _pos < _pattern.Length && _pattern[_pos++] == ')')
+                                if (_capnames != null && _capnames.ContainsKey(capname) && _pos < _pattern.Length && _pattern[_pos++] == ')')
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:989]: Ditto

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs [hunk 9]
```diff
@@ -1002,24 +997,16 @@ private RegexNode ScanReplacement()
...
 
-                        int charsRight = _pattern.Length - _pos;
-                        if (charsRight >= 3 && _pattern[_pos + 1] == '?')
+                        if (_pos + 2 < _pattern.Length && _pattern[_pos + 1] == '?')
                         {
-                            char rightchar2 = _pattern[_pos + 2];
-
                             // disallow comments in the condition
-                            if (rightchar2 == '#')
+                            if (_pattern[_pos + 2] == '#')
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:1003]: The removal of rightchar2 is reintroducing several bounds-checked string accesses that weren't there before.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs [hunk 10]
```diff
@@ -1079,20 +1066,16 @@ private void ScanBlank()
 
                 if ((_options & RegexOptions.IgnorePatternWhitespace) != 0 && _pos < _pattern.Length && _pattern[_pos] == '#')
                 {
-                    while (_pos < _pattern.Length && _pattern[_pos] != '\n')
-                    {
-                        _pos++;
-                    }
+                    _pos = _pattern.IndexOf('\n', _pos);
+                    if (_pos == -1)
+                        _pos = _pattern.Length;
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:1071]:
```C#
if (_pos < 0)
{
    _pos = _pattern.Length;
}
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs [hunk 11]
```diff
@@ -1079,20 +1066,16 @@ private void ScanBlank()
...
                 else if (_pos + 2 < _pattern.Length && _pattern[_pos + 2] == '#' && _pattern[_pos + 1] == '?' && _pattern[_pos] == '(')
                 {
-                    while (_pos < _pattern.Length && _pattern[_pos] != ')')
-                    {
-                        _pos++;
-                    }
-
-                    if (_pos == _pattern.Length)
+                    _pos = _pattern.IndexOf(')', _pos);
+                    if (_pos == -1)
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:1076]:
```C#
if (_pos < 0)
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs [hunk 12]
```diff
@@ -1108,8 +1091,6 @@ private void ScanBlank()
         /// <summary>Scans chars following a '\' (not counting the '\'), and returns a RegexNode for the type of atom scanned</summary>
         private RegexNode? ScanBackslash(bool scanOnly)
         {
-            Debug.Assert(_pos < _pattern.Length, "The current reading position must not be at the end of the pattern");
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:1111]: Why was this removed?

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:1186]: > we don't put asserts in front of every possible null ref.

No, but we do put asserts in places that are meant to act as preconditions / contracts.  These asserts make it really easy for a maintainer to see that these methods should only be called when there's pattern remaining.  I don't think removing them was valuable.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs [hunk 13]
```diff
@@ -1299,7 +1278,7 @@ private void ScanBlank()
                 {
                     return
                         scanOnly ? null :
-                        IsCaptureName(capname) ? new RegexNode(RegexNodeKind.Backreference, _options, CaptureSlotFromName(capname)) :
+                        _capnames != null && _capnames.ContainsKey(capname) ? new RegexNode(RegexNodeKind.Backreference, _options, (int)_capnames![capname]!) :
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:1281]: Same comment as earlier

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs [hunk 14]
```diff
@@ -1395,9 +1374,9 @@ private RegexNode ScanDollar()
                 string capname = ScanCapname();
                 if (_pos < _pattern.Length && _pattern[_pos++] == '}')
                 {
-                    if (IsCaptureName(capname))
+                    if (_capnames != null && _capnames.ContainsKey(capname))
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:1377]: Same comment as earlier

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs [hunk 15]
```diff
@@ -1511,13 +1490,19 @@ private int ScanDecimal()
         private char ScanHex(int c)
         {
             int i = 0;
-            int d;
 
             if (_pos + c <= _pattern.Length)
             {
-                for (; c > 0 && ((d = HexDigit(_pattern[_pos++])) >= 0); c -= 1)
+                for (; c > 0; c -= 1)
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:1496]:
```C#
for (; c > 0; c--)
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs [hunk 16]
```diff
@@ -1511,13 +1490,19 @@ private int ScanDecimal()
...
                 {
-                    i = (i * 0x10) + d;
+                    int d;
+                    char ch = _pattern[_pos++];
+                    if ((uint)(d = ch - '0') <= 9)
+                        i = (i * 0x10) + d;
+                    else if ((uint)(d = (ch | 0x20) - 'a') <= 5)
+                        i = (i * 0x10) + d + 0xa;
+                    else
+                        break;
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:1505]:
```C#
                    if ((uint)(d = ch - '0') <= 9)
                    {
                        i = (i * 0x10) + d;
                    }
                    else if ((uint)(d = (ch | 0x20) - 'a') <= 5)
                    {
                        i = (i * 0x10) + d + 0xa;
                    }
                    else
                    {
                        break;
                    }
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs [hunk 17]
```diff
@@ -1511,13 +1490,19 @@ private int ScanDecimal()
...
-            int d;
 
             if (_pos + c <= _pattern.Length)
             {
-                for (; c > 0 && ((d = HexDigit(_pattern[_pos++])) >= 0); c -= 1)
+                for (; c > 0; c -= 1)
                 {
-                    i = (i * 0x10) + d;
+                    int d;
+                    char ch = _pattern[_pos++];
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:1499]: Can this use our HexConverter internal utility instead?

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs (hunk 10 above)
Comment (danmoseley) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:1071]: Cheaper?

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs (hunk 12 above)
Comment (danmoseley) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:1111]: It will throw out of bounds just below. We generally don't assert indexes just before we use them.

Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:1111]: We do generally assert preconditions.  Calling this when there's pattern remaining is a precondition.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs (hunk 17 above)
Comment (danmoseley) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:1499]: I looked and it seemed a lot more indirection that surely is more expensive than this is given we're just converting two digits. But I can look again if you like.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs (hunk 10 above)
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:1071]: A bit, and also matches our style preferences and the style of the rest of the code better.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs (hunk 17 above)
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:1499]: > I looked and it seemed a lot more indirection that surely is more expensive than this is given we're just converting two digits

What indirection are you referring to?  I'm referring to https://github.com/dotnet/runtime/blob/b898a540f75d78ca55ee33288a4fd62b1ab69fa2/src/libraries/Common/src/System/HexConverter.cs#L357-L361. It uses a lookup table.  If it were more expensive then the multiple checks being performed here, then that centralized implementation would be changed to the more efficient thing.

**Comment by stephentoub** [comment]
> I ran the real world patterns corpus through the parser and didn't notice anything that could be obviously improved except for one dictionary which was almost always resizing. Set size to eliminate resizing in 90% of cases.

I'm not seeing it... which line is the change to presize the dictionary?

**Comment by danmoseley** [comment]
> I'm not seeing it... which line is the change to presize the dictionary?

Thanks for spotting that -- it got lost in a rebase, apparently:
https://github.com/dotnet/runtime/compare/main...danmoseley:regex.presize?expand=1

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs (hunk 9 above)
Comment (danmoseley) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:1003]: the jit appears to be [smart enough](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgAYBYAKDIEYabi6BOACgCJAeDcBB9jgSgDcjWnSQACYinEAxNmAAW2KAG0AuuID6AB2wYMMKADs04gJZGMW7RFz8aAbxHiX5y+MXLcAJTMBzBSsAXms9A2MAOgAZGCM/DAVxOGtbYWpXcwAzcXklKB9/QPEAPhCAZnEAMkrQ/UMjFR1bcQBqcToNIJCAcgB+bvt01ychjJdPKHEoQowJ0nEQnTD6xptcVvFSNTSaMdcAen3xABMzXGwAGwuIAHcPCHx8WIx1i3EEmHujU4wzCCNdnssjlpgFZnl5l1xN0AMQDQF7EZAsYJKC3cRGGB3ACiCDAMG0v3+bCECLGAF9nEDDiczpdrncjNgnscPNhCQBXWDiPxojnaHK9AA8EQixVF/Dc7wUn0g3zMRIBozGZmybFBgTmCx6AB1uvDlRkkcjXKj0ZicXiCYqSWlkZTqGSMqqchMCmCSiFpNUpjMtVDukLulUaks6sZVs02mUNABCHqx4M+sPhBpNdbRuM9IIG5HGk3StGMrHiXH4wl/Iy2p2uB0ZB0OkTMCRSWSkXLKdS1VOmCxWdODfNuKxu3wexa6cNGaKxeKJZLpu2uF0d/JjoqlcQVZOT1ORjPtTo9fqDI01lw007nK7oyCPZ6vIzS2X/H6V8/AtgplbpjZbbXQnCp5AkOyJmsWlrljapKGrWVJ7JedI3oyzIwKyYDshgXKfLyED8oKIpihKUofF8b7/B+K7fhGv5tP+AZ6rmIEfhk4EYiWZbWpW1awS4ja8Z+o4zJ64jeqGu4/msf5HtCQYht2klRluWbQom8nUWmUmZuI8bQjmwGIixpoKEW7GQVxxIwfaZINkAA)?

Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:1003]: > the jit appears to be smart enough

_pattern and _pos are fields, not parameters

**Comment by stephentoub** [comment]
>> Thanks for spotting that -- it got lost in a rebase, apparently: https://github.com/dotnet/runtime/compare/main...danmoseley:regex.presize?expand=1

> _stringTable = new Dictionary<string, int>(15); // avoids resize for 90% of real world patterns

How much of an overallocation is it for those? i.e. we could probably avoid it for 100% of the real-world patterns if we made it `new Dictionary<string, int>(1_000_000)`, but that's obviously a bad answer.

Said another way, what does the histogram look like for number of strings involved in the various patterns?

**Comment by danmoseley** [comment]
https://gist.github.com/danmoseley/eeb38412d74c3eb22bc69472940b1f95

image

Note horizontal scale starts skipping around #60

**Comment by danmoseley** [comment]
If the dictionary starts at zero, resizes when full to the next prime at least 2x the size, then it will have sizes 0, 3, 7, 17, 37, 89

61% would fit in 3, 80% in 7 and 96% in 17. Only 5% are zero and would not allocate at all.

| 0  | 12647 | 0.05 | 3  |
| -- | ----- | ---- | -- |
| 1  | 62735 | 0.31 | 3  |
| 2  | 43475 | 0.50 | 3  |
| 3  | 27852 | 0.61 | 3  |
| 4  | 17408 | 0.69 | 7  |
| 5  | 11870 | 0.74 | 7  |
| 6  | 8788  | 0.77 | 7  |
| 7  | 7623  | 0.80 | 7  |
| 8  | 7035  | 0.83 | 17 |
| 9  | 6381  | 0.86 | 17 |
| 10 | 5069  | 0.88 | 17 |
| 11 | 4178  | 0.90 | 17 |
| 12 | 3397  | 0.91 | 17 |
| 13 | 3023  | 0.93 | 17 |
| 14 | 2840  | 0.94 | 17 |
| 15 | 2243  | 0.95 | 17 |
| 16 | 2103  | 0.96 | 17 |
| 17 | 1800  | 0.96 | 17 |
| 18 | 1169  | 0.97 | 37 |
| 19 | 1360  | 0.97 | 37 |
| 20 | 1157  | 0.98 | 37 |
| 21 | 925   | 0.98 | 37 |
| 22 | 828   | 0.99 | 37 |
| 23 | 468   | 0.99 | 37 |
| 24 | 425   | 0.99 | 37 |
| 25 | 314   | 0.99 | 37 |
| 26 | 264   | 0.99 | 37 |
| 27 | 240   | 0.99 | 37 |
| 28 | 293   | 0.99 | 37 |
| 29 | 188   | 0.99 | 37 |
| 30 | 38    | 0.99 | 37 |
| 31 | 86    | 1.00 | 37 |

**Comment by stephentoub** [comment]
> 61% would fit in 3, 80% in 7 and 96% in 17. Only 5% are zero and would not allocate at all.

Right.  So presizing to 15 would overallocate for the majority.

**Comment by danmoseley** [comment]
I guess it depends how you trade off a small transient allocation with a small CPU cost of copying. I do not think this is a hot path when aggregated with actually running the interpreter. But it seemed that we have some hard data here and it was easy to avoid that CPU cost in 90% of cases. I don't mind either way.

**Comment by stephentoub** [comment]
Thanks.

> But it seemed that we have some hard data here and it was easy to avoid that CPU cost in 90% of cases.

There's CPU cost associated with overallocation as well.

I don't think presizing it to 15 is the right tradeoff.
